New gene translation efficiency modification

rjuenemann opened the pull request:
This PR builds upon #1350 and enables the manipulation of the translation efficiency of new genes through a variant, which provides an additional dimension for investigating the source of cellular burden when new genes are overexpressed. The new variant had to include both new gene expression and translation efficiency because using multiple variants within the same set of simulations is not possible at this time.

A limitation was added to the polypeptide initiation process to ensure physical limitations are not violated with respect to ribosome footprint size. This was modeled after a similar limitation previously added by @ggsun for RNA polymerases in #1325. Analysis scripts were added to compare the target translation probabilities and actual translation probabilities for mRNAs whose translation probabilities exceeded the limit set by the physical size and the elongation rates of ribosomes.

I attached an example output of each new analysis script. The simulation results depicted in this data are for the addition of a single gfp gene at multiple expression and translation efficiency levels, run with multiple replicates (seeds). I have chosen to include the plots corresponding to later generations (4-15) and filtered out the cells that did not divide before the max time threshold (both can be changed via flags at the beginning of each analysis script), as this is most similar to our analysis of these simulations. 

For additional details on how to use this additional input to new gene functionality, please refer to docs/misc/add_new_gene.md

[new_gene_exp_trl_eff_dashboard_late_gens.pdf](https://github.com/CovertLab/wcEcoli/files/11887371/new_gene_exp_trl_eff_dashboard_late_gens.pdf)
[single_ribosome_crowding.pdf](https://github.com/CovertLab/wcEcoli/files/11887372/single_ribosome_crowding.pdf)
[multigen_ribosome_crowding.pdf](https://github.com/CovertLab/wcEcoli/files/11887369/multigen_ribosome_crowding.pdf)
[multigen_rnap_crowding.pdf](https://github.com/CovertLab/wcEcoli/files/11887370/multigen_rnap_crowding.pdf)
[variant_ribosome_crowding_late_gens.txt](https://github.com/CovertLab/wcEcoli/files/11887374/variant_ribosome_crowding_late_gens.txt)
[variant_rnap_crowding_late_gens.txt](https://github.com/CovertLab/wcEcoli/files/11887375/variant_rnap_crowding_late_gens.txt)

ggsun commented on models/ecoli/analysis/variant/new_gene_translation_efficiency_heatmaps.py:
This analysis script is pretty difficult to read for someone who's never used it before, and I'm a bit concerned about how easy it would be to expand this in the future. You'd probably have a better understanding of how it works and how to add more heatmaps, but I think it might be difficult for you as well if you return to this script after some time. I might suggest trying to remove some of the lesser-used options (do we expect to want to see the early gen heatmaps often?) if it makes the code simpler, and maybe modularizing the heatmap building a bit further, even if we have to sacrifice some formatting flexibility for each heatmap. Considering this is a dashboard plot, and not really intended as a publication-grade plot, I wouldn't be too concerned about formatting and optimize for ease of use and extensibility.

ggsun commented on models/ecoli/analysis/variant/ribosome_crowding.py:
```diff
@@ -0,0 +1,168 @@
+from wholecell.analysis.analysis_tools import read_stacked_columns, \
+	stacked_cell_threshold_mask
+
+# 1 to exclude cells that took full MAX_CELL_LENGTH, 0 otherwise
+exclude_timeout_cells = 1
+
+"""
+1 to plot early (before MIN_LATE_CELL_INDEX), and late generations in
+addition to all generations
+"""
+exclude_early_gens = 1
```
Same comment as before - boolean and all caps

ggsun commented on models/ecoli/analysis/variant/ribosome_crowding.py:
```diff
@@ -0,0 +1,168 @@
+"""
+MIN_LATE_CELL_INDEX = 4
+
+MAX_CELL_LENGTH = 36000
```
Might be good to specify what units this is as a comment - I'm assuming this is in seconds?

ggsun commented on models/ecoli/analysis/variant/ribosome_crowding.py:
```diff
@@ -0,0 +1,168 @@
+				for monomer_id in overcrowded_monomer_index_dicts[j]:
+					overcrowded_variant_string = ", ".join(map(str,
+						overcrowded_monomer_index_dicts[j][monomer_id]))
+					f.write("" + overcrowded_gene_indexes_to_ids_dict[
+						monomer_id] + ": " + overcrowded_variant_string + "\n")
```
Using [f-strings](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/) might be a more cleaner/flexible option for formatting these types of strings.

ggsun commented:
I haven't looked at some of the changed files yet, I'll try to leave more comments later this week!

ggsun commented on docs/misc/add_new_gene.md:
```diff
@@ -29,35 +29,91 @@ The following steps occur in incorporating the new genes into the chromosome:
-* Listeners (`models/ecoli/listeners/mRNA_counts.py` and `models/ecoli/listeners/protein_counts.py`) did not have to be modified
+<b>Listeners</b><br>
+* Listeners `models/ecoli/listeners/mRNA_counts.py` and `models/ecoli/listeners/protein_counts.py` did not have to be modified
+* Listener `models/ecoli/listeners/ribisome_data.py` and process 
```
There's a typo in `ribisome_data.py`.

rjuenemann commented:
Hi @ggsun,

Thanks for your feedback! I incorporated the changes, and reworked `models/ecoli/analysis/variant/new_gene_translation_efficiency_heatmaps.py.` The reworked version had a similar runtime on Sherlock. Could you take a look at this analysis script and let me know what you think? 

Thank you!

rjuenemann commented:
Discussed in person - merging in!
